File: backend/src/services/submission.js
const uuid = require('uuid');
const db = require('../db/submission');
const { submissionStatus } = require('../consts');
const { createJobRun } = require('./outbound/jobExecutor');

const createSubmission = submission => {
  const s = submission;
  s.id = uuid.v4();
  s.result = '{}';
  s.submissionStatusName = submissionStatus.PENDING;
  return db.createSubmission(submission);
};

const createSubmissionJob = submissionId => createJobRun(submissionId);

const getSubmissions = (email, assignmentId) => db.getSubmissions(email, assignmentId);

const getSubmission = (id, email) => db.getSubmission(id, email);

module.exports = {
  createSubmission,
  createSubmissionJob,
  getSubmissions,
  getSubmission
};

File: pkg/controller/submission.go
package controller

import (
	"context"

	"github.com/ZdravkoGyurov/grader/pkg/types"
	"github.com/google/uuid"
)

func (c *Controller) CreateSubmission(ctx context.Context, submission *types.Submission) (*types.Submission, error) {
	submission.ID = uuid.NewString()
	submission.Result = ""
	submission.SubmissionStatusName = types.SubmissionStatusPending

	if err := submission.ValidateCreate(); err != nil {
		return nil, err
	}

	if err := c.storage.CreateSubmission(ctx, submission); err != nil {
		return nil, err
	}

	return submission, nil
}

func (c *Controller) CreateSubmissionJob(ctx context.Context, submissionID string) error {
	return c.createJobRun(ctx, submissionID)
}

func (c *Controller) GetSubmissions(ctx context.Context, userEmail, assignmentID string) ([]*types.Submission, error) {
	return c.storage.GetSubmissions(ctx, userEmail, assignmentID)
}

func (c *Controller) GetSubmission(ctx context.Context, id, userEmail string) (*types.Submission, error) {
	return c.storage.GetSubmission(ctx, id, userEmail)
}
